frb_rust/src/ffi/web.rs

```diff
@@ -0,0 +1,388 @@
+    ($( $ty:ty )*) => {$(
+        impl IntoDart for $ty {
+            #[inline]
+            fn into_dart(self) -> DartAbi {
+                DartAbi::from(self)
+            }
+        }
+    )*};
+}
+macro_rules! delegate_buffer {
+    ($( $ty:ty => $buffer:ty )*) => {$(
+        impl IntoDart for $ty {
+            #[inline]
+            fn into_dart(self) -> DartAbi {
+                <$buffer>::from(self.as_slice()).into()
+            }
+        }
+    )*};
+}
+// Orphan rules disallow blanket implementations, so we have to manually delegate here.
+delegate! {
+    bool
+    i8 u8 i16 u16 i32 u32 i64 u64 i128 u128 isize usize
+    f32 f64
+    &str String JsValue
+}
+delegate_buffer! {
+    Vec<i8> => js_sys::Int8Array
+    Vec<u8> => js_sys::Uint8Array
+    Vec<i16> => js_sys::Int16Array
+    Vec<u16> => js_sys::Uint16Array
+    Vec<i32> => js_sys::Int32Array
+    Vec<u32> => js_sys::Uint32Array
+    Vec<f32> => js_sys::Float32Array
+    Vec<f64> => js_sys::Float64Array
+    ZeroCopyBuffer<Vec<i8>> => js_sys::Int8Array
+    ZeroCopyBuffer<Vec<u8>> => js_sys::Uint8Array
+    ZeroCopyBuffer<Vec<i16>> => js_sys::Int16Array
+    ZeroCopyBuffer<Vec<u16>> => js_sys::Uint16Array
+    ZeroCopyBuffer<Vec<i32>> => js_sys::Int32Array
+    ZeroCopyBuffer<Vec<u32>> => js_sys::Uint32Array
+    ZeroCopyBuffer<Vec<f32>> => js_sys::Float32Array
+    ZeroCopyBuffer<Vec<f64>> => js_sys::Float64Array
+}
+
+impl<T: IntoDartExceptPrimitive> IntoDart for Vec<T> {
+    #[inline]
+    fn into_dart(self) -> DartAbi {
+        Array::from_iter(self.into_iter().map(IntoDart::into_dart)).into()
+    }
+}
+
+impl<T: IntoDart> IntoDart for Option<T> {
+    #[inline]
+    fn into_dart(self) -> DartAbi {
+        self.map(T::into_dart).unwrap_or_else(JsValue::null)
+    }
+}
+impl<T> IntoDart for *const T {
+    #[inline]
+    fn into_dart(self) -> DartAbi {
+        (self as u32).into_dart()
+    }
+}
+impl<T> IntoDart for *mut T {
+    #[inline]
+    fn into_dart(self) -> DartAbi {
+        (self as u32).into_dart()
+    }
+}
+
+impl<const N: usize, T: IntoDartExceptPrimitive> IntoDart for [T; N] {
+    #[inline]
+    fn into_dart(self) -> DartAbi {
+        let boxed: Box<[T]> = Box::new(self);
+        boxed.into_vec().into_dart()
+    }
+}
+
+impl<const N: usize> IntoDart for [u8; N] {
+    #[inline]
+    fn into_dart(self) -> DartAbi {
+        Vec::from(self).into_dart()
+    }
+}
+
+macro_rules! delegate_big_buffers {
+    ($buf:ident, $prim:ty => $buffer:ty) => {
+        const RATIO: usize = core::mem::size_of::<$prim>() / core::mem::size_of::<i32>();
+        // Multiply the length by the difference in byte length
+        let len = $buf.len() * RATIO;
+        // Turn Vec into a Box, dropping excess capacity.
+        let ptr = Box::into_raw($buf.into_boxed_slice()) as *mut i32;
+        // Reassemble into a new slice, now with longer length and shorter byte length
+        let buf = unsafe { Box::from_raw(core::slice::from_raw_parts_mut(ptr, len)) };
```

Comment:
If I understand correctly, these part is doing one thing: Convert `Vec<i64>` to `Box<[i32]>` by making one i64 two i32s?

If so, what about this well-tested crate that wraps unsafe into safe operations: https://docs.rs/byte-slice-cast/latest/byte_slice_cast/ and/or https://docs.rs/bytemuck/latest/bytemuck/. Then we avoid unsafe operations!

Reply:
Oh yeah, I forgot that I already added bytemuck as a dependency but didn't use it.